File: python_port/core/logic/battle/map_unit_combatant.py
from typing import Sequence
from com.unciv.logic.civilization import Civilization
from com.unciv.logic.map.mapunit import MapUnit
from com.unciv.logic.map.tile import Tile
from com.unciv.models import UncivSound
from com.unciv.models.ruleset.unique import StateForConditionals, Unique, UniqueType
from com.unciv.models.ruleset.unit import UnitType
from com.unciv.logic.battle.i_combatant import ICombatant

class MapUnitCombatant(ICombatant):
    def __init__(self, unit: MapUnit):
        self.unit = unit

    def get_health(self) -> int:
        return self.unit.health

    def get_max_health(self) -> int:
        return 100

    def get_civ_info(self) -> Civilization:
        return self.unit.civ

    def get_tile(self) -> Tile:
        return self.unit.get_tile()

    def get_name(self) -> str:
        return self.unit.name

    def is_defeated(self) -> bool:
        return self.unit.health <= 0

    def is_invisible(self, to: Civilization) -> bool:
        return self.unit.is_invisible(to)

    def can_attack(self) -> bool:
        return self.unit.can_attack()

    def matches_filter(self, filter: str, multi_filter: bool = True) -> bool:
        return self.unit.matches_filter(filter, multi_filter)

    def get_attack_sound(self) -> UncivSound:
        sound = self.unit.base_unit.attack_sound
        return UncivSound.Click if sound is None else UncivSound(sound)

    def take_damage(self, damage: int) -> None:
        self.unit.take_damage(damage)

    def get_attacking_strength(self) -> int:
        return self.unit.base_unit.ranged_strength if self.is_ranged() else self.unit.base_unit.strength

    def get_defending_strength(self, attacked_by_ranged: bool = False) -> int:
        if self.unit.is_embarked() and not self.is_civilian():
            return self.unit.civ.get_era().embark_defense
        elif self.is_ranged() and attacked_by_ranged:
            return self.unit.base_unit.ranged_strength
        return self.unit.base_unit.strength

    def get_unit_type(self) -> UnitType:
        return self.unit.type

    def __str__(self) -> str:
        return f"{self.unit.name} of {self.unit.civ.civ_name}"

    def get_matching_uniques(self, unique_type: UniqueType, conditional_state: StateForConditionals, check_civ_uniques: bool) -> Sequence[Unique]:
        return self.unit.get_matching_uniques(unique_type, conditional_state, check_civ_uniques)

    def has_unique(self, unique_type: UniqueType, conditional_state: StateForConditionals | None = None) -> bool:
        if conditional_state is None:
            return self.unit.has_unique(unique_type)
        return self.unit.has_unique(unique_type, conditional_state)

File: unciv-rust-port/src/battle/map_unit_combatant.rs
use crate::battle::i_combatant::ICombatant;
use crate::civilization::Civilization;
use crate::map::map_unit::MapUnit;
use crate::map::tile::Tile;
use crate::models::unciv_sound::UncivSound;
use crate::models::ruleset::unique::state_for_conditionals::StateForConditionals;
use crate::models::ruleset::unique::unique::Unique;
use crate::models::ruleset::unique::unique_type::UniqueType;
use crate::models::unit_type::UnitType;
use std::sync::Arc;
use std::fmt;

/// Represents a map unit as a combatant in battle
#[derive(Clone)]
pub struct MapUnitCombatant {
    /// The unit this combatant represents
    pub unit: Arc<MapUnit>,
}

impl std::fmt::Debug for MapUnitCombatant {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("MapUnitCombatant")
            .field("unit", &format!("Arc<MapUnit>({:p})", Arc::as_ptr(&self.unit)))
            .finish()
    }
}

impl MapUnitCombatant {
    /// Creates a new MapUnitCombatant from a MapUnit
    pub fn new(unit: Arc<MapUnit>) -> Self {
        MapUnitCombatant { unit }
    }

    /// Gets matching uniques for the unit
    pub fn get_matching_uniques(
        &self,
        unique_type: UniqueType,
        conditional_state: &StateForConditionals,
        check_civ_uniques: bool
    ) -> Vec<Unique> {
        self.unit.get_matching_uniques(unique_type, conditional_state, check_civ_uniques)
    }

    /// Checks if the unit has a specific unique
    pub fn has_unique(
        &self,
        unique_type: UniqueType,
        conditional_state: Option<&StateForConditionals>
    ) -> bool {
        match conditional_state {
            Some(state) => self.unit.has_unique(unique_type, state),
            None => self.unit.has_unique(unique_type),
        }
    }
}

impl ICombatant for MapUnitCombatant {
    fn get_health(&self) -> i32 {
        self.unit.health
    }

    fn get_max_health(&self) -> i32 {
        100
    }

    fn get_civ_info(&self) -> Arc<Civilization> {
        self.unit.civ.clone()
    }

    fn get_tile(&self) -> Arc<Tile> {
        self.unit.get_tile()
    }

    fn get_name(&self) -> String {
        self.unit.name.clone()
    }

    fn is_defeated(&self) -> bool {
        self.unit.health <= 0
    }

    fn is_invisible(&self, to: &Civilization) -> bool {
        self.unit.is_invisible(to)
    }

    fn can_attack(&self) -> bool {
        self.unit.can_attack()
    }

    fn matches_filter(&self, filter: &str, multi_filter: bool) -> bool {
        self.unit.matches_filter(filter, multi_filter)
    }

    fn get_attack_sound(&self) -> UncivSound {
        match &self.unit.base_unit.attack_sound {
            Some(sound) => UncivSound::from(sound.clone()),
            None => UncivSound::Click,
        }
    }

    fn take_damage(&mut self, damage: i32) {
        self.unit.take_damage(damage);
    }

    fn get_attacking_strength(&self) -> i32 {
        if self.is_ranged() {
            self.unit.base_unit.ranged_strength
        } else {
            self.unit.base_unit.strength
        }
    }

    fn get_defending_strength(&self, attacked_by_ranged: bool) -> i32 {
        if self.unit.is_embarked() && !self.is_civilian() {
            self.unit.civ.get_era().embark_defense
        } else if self.is_ranged() && attacked_by_ranged {
            self.unit.base_unit.ranged_strength
        } else {
            self.unit.base_unit.strength
        }
    }

    fn get_unit_type(&self) -> UnitType {
        self.unit.type_.clone()
    }

    fn is_ranged(&self) -> bool {
        self.unit.base_unit.is_ranged()
    }

    fn is_air_unit(&self) -> bool {
        self.unit.base_unit.is_air_unit()
    }

    fn is_water_unit(&self) -> bool {
        self.unit.base_unit.is_water_unit
    }

    fn is_land_unit(&self) -> bool {
        self.unit.base_unit.is_land_unit
    }

    fn is_city(&self) -> bool {
        false
    }

    fn is_civilian(&self) -> bool {
        self.unit.is_civilian()
    }
}

impl fmt::Display for MapUnitCombatant {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} of {}", self.unit.name, self.unit.civ.civ_name)
    }
}